libraries/persistent_store/src/concat.rs

```diff
@@ -0,0 +1,242 @@
+        Some(x) => x,
+    };
+    Ok(find(&values, index)?.map(|range| values[range].to_vec()))
+}
+
+/// Writes a value to a concatenated entry.
+pub fn write(
+    store: &mut Store<impl Storage>,
+    key: usize,
+    index: u8,
+    value: &[u8],
+) -> StoreResult<()> {
+    if value.len() > 255 {
+        return Err(StoreError::InvalidArgument);
+    }
+    let mut values = store.find(key)?.unwrap_or(vec![]);
+    match find(&values, index)? {
+        None => {
+            values.push(index);
+            values.push(value.len() as u8);
+            values.extend_from_slice(value);
+        }
+        Some(mut range) => {
+            values[range.start - 1] = value.len() as u8;
+            match range.len().cmp(&value.len()) {
+                Ordering::Less => {
+                    let diff = value.len() - range.len();
+                    values.resize(values.len() + diff, 0);
+                    values[range.end..].rotate_right(diff);
+                    range.end += diff;
+                }
+                Ordering::Equal => (),
+                Ordering::Greater => {
+                    let diff = range.len() - value.len();
+                    range.end -= diff;
+                    values[range.end..].rotate_left(diff);
+                    values.truncate(values.len() - diff);
+                }
+            }
+            values[range].copy_from_slice(value);
+        }
+    }
+    store.insert(key, &values)
+}
+
+/// Deletes the value from a concatenated entry.
+pub fn delete(store: &mut Store<impl Storage>, key: usize, index: u8) -> StoreResult<()> {
+    let mut values = match store.find(key)? {
+        None => return Ok(()),
+        Some(x) => x,
+    };
+    let mut range = match find(&values, index)? {
+        None => return Ok(()),
+        Some(x) => x,
+    };
+    range.start -= 2;
+    values[range.start..].rotate_left(range.len());
+    values.truncate(values.len() - range.len());
+    store.insert(key, &values)
+}
+
+fn find(values: &[u8], index: u8) -> StoreResult<Option<Range<usize>>> {
+    let mut pos = 0;
+    while pos < values.len() {
+        if pos == values.len() - 1 {
+            return Err(StoreError::InvalidStorage);
+        }
+        let len = values[pos + 1] as usize;
+        if len > values.len() - 2 || pos > values.len() - 2 - len {
+            return Err(StoreError::InvalidStorage);
+        }
+        if index == values[pos] {
+            return Ok(Some(pos + 2..pos + 2 + len));
+        }
+        pos += 2 + len;
+    }
+    Ok(None)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test::MINIMAL;
+
+    #[test]
+    fn read_empty_entry() {
+        let store = MINIMAL.new_store();
+        assert_eq!(read(&store, 0, 0), Ok(None));
+        assert_eq!(read(&store, 0, 1), Ok(None));
+    }
+
+    #[test]
+    fn read_missing_value() {
+        let mut store = MINIMAL.new_store();
+        let value = b"\x00\x03foo\x02\x05hello".to_vec();
```

Comment:
The tests are arguably more maintainable if they insert into the store through `concat`'s `write`, in case we'd ever change the implementation. But we have to write arrays explicitly for the invalid storage tests, so works for me.

Reply:
Yes I think that's arguable. Those tests are really unit tests (they test a single function) rather than functional tests (they test global behavior). Because the implementation is part of the specification, testing the implementation is actually testing the specification. Also, I don't think we should change the implementation because that would be a breaking change, which is why the implementation is specified.

Reply:
Okay, SGTM.